pandas/core/sparse/array.py

```diff
@@ -120,298 +135,443 @@ def _wrap_result(name, data, sparse_index, fill_value, dtype=None):
+        )
+        self._sparse_index = sparse_index
+        self._sparse_values = sparse_values
+        self._dtype = dtype
+        return self
 
-    def __array_wrap__(self, out_arr, context=None):
+    @property
+    def kind(self):
         """
-        NumPy calls this method when ufunc is applied
+        The kind of sparse index for this array. One of {'integer', 'block'}.
+        """
+        # TODO: make this an abstract attribute of SparseIndex
+        if isinstance(self.sp_index, IntIndex):
+            return 'integer'
+        else:
+            return 'block'
 
-        Parameters
-        ----------
+    @property
+    def _valid_sp_values(self):
+        sp_vals = self.sp_values
+        mask = notna(sp_vals)
+        return sp_vals[mask]
 
-        out_arr : ndarray
-            ufunc result (note that ufunc is only applied to sp_values)
-        context : tuple of 3 elements (ufunc, signature, domain)
-            for example, following is a context when np.sin is applied to
-            SparseArray,
+    def __len__(self):
+        return self.sp_index.length
 
-            (<ufunc 'sin'>, (SparseArray,), 0))
+    @property
+    def _null_fill_value(self):
+        return self._dtype._is_na_fill_value
 
-        See http://docs.scipy.org/doc/numpy/user/basics.subclassing.html
-        """
-        if isinstance(context, tuple) and len(context) == 3:
-            ufunc, args, domain = context
-            # to apply ufunc only to fill_value (to avoid recursive call)
-            args = [getattr(a, 'fill_value', a) for a in args]
-            with np.errstate(all='ignore'):
-                fill_value = ufunc(self.fill_value, *args[1:])
+    def _fill_value_matches(self, fill_value):
+        if self._null_fill_value:
+            return pd.isna(fill_value)
         else:
-            fill_value = self.fill_value
+            return self.fill_value == fill_value
 
-        return self._simple_new(out_arr, sp_index=self.sp_index,
-                                fill_value=fill_value)
+    @property
+    def nbytes(self):
+        return self.sp_values.nbytes + self.sp_index.nbytes
 
-    def __array_finalize__(self, obj):
+    @property
+    def values(self):
         """
-        Gets called after any ufunc or other array operations, necessary
-        to pass on the index.
+        Dense values
         """
-        self.sp_index = getattr(obj, 'sp_index', None)
-        self._fill_value = getattr(obj, 'fill_value', None)
+        return self.to_dense()
 
-    def __reduce__(self):
-        """Necessary for making this object picklable"""
-        object_state = list(np.ndarray.__reduce__(self))
-        subclass_state = self.fill_value, self.sp_index
-        object_state[2] = self.sp_values.__reduce__()[2]
-        object_state[2] = (object_state[2], subclass_state)
-        return tuple(object_state)
+    def isna(self):
+        # Two unfortunate things here:
+        # 1. We can't
```

Comment:
Turned this into a test :) https://github.com/pandas-dev/pandas/pull/22325/files#diff-973da69b224ebf24c470136a6f73fb22R981

The basic issue that setting `SparseArray.fill_value` has the potential to ruin everything. Several methods (before and after this refactor) assume that `fill_value` doesn't show up in `self.sp_values`. Assuming that let's us take shortcuts (e.g. fillna, we would know that if we have an NA fill value then sp_values are all valid).

The second issue is that this will return an in-memory ndarray. We could return `Sparse[bool]`, with the same sparsity as `self`, but that's liable to break other things.

Reply:
Hmm, it would be easier if for such things the fill (empty) values would simply not be considered .. (so we would only need to isna the `sp_values` (and then returning a `Sparse[bool]` also makes much more sense). But I suppose that's quite a deviation in sparse model?

Reply:
I still think `isna` should not need to return a dense boolean array? 
I would suppose it is very similar to boolean operators like `arr > 1`, that still return a boolean sparse array where the fill value is True or False depending on the comparison of the fill value